File: projeto_ed_para_praticar/src/main/java/repository/dinamicas/pilha/PilhaDinamica.java
package repository.dinamicas.pilha;

import java.util.NoSuchElementException;

import repository.Empilhavel;
import repository.ed_dinamicas.NoDuplo;

public class PilhaDinamica implements Empilhavel {
    private NoDuplo ponteiroTopo;
    private int tamanhoMaximo;
    private int quantidade;

    public PilhaDinamica() {
        this(10);
    }

    public PilhaDinamica(int tamanhoMaximo) {
        ponteiroTopo = null;
        this.tamanhoMaximo = tamanhoMaximo;
        quantidade = 0;
    }

    @Override
    public Object desempilhar() {
        if (estaVazia()) {
            throw new NoSuchElementException("Pilha Vazia!");
        }
        Object dadoTopo = ponteiroTopo.getDado();
        ponteiroTopo = ponteiroTopo.getAnterior();
        if (ponteiroTopo != null) {
            ponteiroTopo.setProximo(null);
        }
        quantidade--;
        return dadoTopo;
    }

    @Override
    public void empilhar(Object dado) {
        if (estaCheia()) {
            throw new NoSuchElementException("Pilha Cheia!");
        }
        NoDuplo novoNo = new NoDuplo(dado);
        if (ponteiroTopo == null) {
            ponteiroTopo = novoNo;
        } else {
            novoNo.setAnterior(ponteiroTopo);
            ponteiroTopo.setProximo(novoNo);
            ponteiroTopo = novoNo;
        }
        quantidade++;
    }

    @Override
    public void atualizar(Object dado) {
        if (estaVazia()) {
            throw new NoSuchElementException("Pilha Vazia!");
        }
        ponteiroTopo.setDado(dado);
    }

    @Override
    public Object espiar() {
        if (estaVazia()) {
            throw new NoSuchElementException("Pilha Vazia!");
        }
        return ponteiroTopo.getDado();
    }

    @Override
    public boolean estaCheia() {
        return quantidade == tamanhoMaximo;
    }

    @Override
    public boolean estaVazia() {
        return quantidade == 0;
    }

    @Override
    public String imprimir() {
        String resultado = "";
        NoDuplo ponteiroAuxiliar = ponteiroTopo;
        for (int i = 0; i < quantidade; i++) {
            resultado += ponteiroAuxiliar.getDado();
            if (i != quantidade - 1) {
                resultado += ",";
            }
            ponteiroAuxiliar = ponteiroAuxiliar.getAnterior();
        }
        return "[" + resultado + "]";
    }

}

File: pilha/Pilha Dinâmica/Kotlin/PilhaDinamica.kt
class PilhaDinamica(private val tamanho: Int = 10) : Empilhavel {

	private var ponteiroTopo: NoDuplo? = NoDuplo()
	private var quantidade: Int = 0
 
	override fun empilhar(dado: Any?) {
		if (!estaCheia()) {
			var noTemp: NoDuplo? = NoDuplo(dado)
			//noTemp?.dado = dado
			noTemp?.anterior = ponteiroTopo
			if (!estaVazia()) {
				ponteiroTopo?.proximo = noTemp
			}
			ponteiroTopo = noTemp
			quantidade++
		} else {
			println("Pilha Cheia!")
		}
	}
	
	override fun desempilhar(): Any? {
		var dadoTopo: Any? = null
		if (!estaVazia()) {
			dadoTopo = ponteiroTopo?.dado
			ponteiroTopo = ponteiroTopo?.anterior
			quantidade--
			if (!estaVazia()) {
				ponteiroTopo?.proximo = null
			}
		} else {
			println("Pilha Vazia!")
		}
		return dadoTopo
	}
	
	override fun espiar(): Any? {
		var dadoTopo: Any? = null
		if (!estaVazia()) {
			dadoTopo = ponteiroTopo?.dado
		} else {
			println("Pilha Vazia!")
		}
		return dadoTopo
	}
	
	override fun atualizar(dado: Any?) {
		if (!estaVazia()) {
			ponteiroTopo?.dado = dado
		} else {
			println("Pilha Vazia!")
		}
	}

	override fun estaCheia(): Boolean {
		return quantidade == tamanho
	}
	
	override fun estaVazia(): Boolean {
		return quantidade == 0
	}
	
	override fun imprimir(): String {
		var ponteiroAuxiliar: NoDuplo? = ponteiroTopo
		var resultado : String = "["
		for (i in 0 until quantidade) {
			if (i == quantidade-1)
				resultado += "${ponteiroAuxiliar?.dado}"
			else
				resultado += "${ponteiroAuxiliar?.dado},\n"
			
			ponteiroAuxiliar = ponteiroAuxiliar?.anterior
		}
		return (resultado + "]")
	}
}